### ISSUE TITLE
Adding IF NOT EXISTS in createTable() method

### ISSUE THREAD

**Issue by pmoust** [yiisoft/yii#3148]
On https://github.com/yiisoft/yii/blob/1.1.14/framework/db/schema/CDbSchema.php#L370 I believe there should be either an option to pass to add `IF NOT EXIST` or always have it in.

I just experienced an issue when setting migration mark to a past/earlier migration, and could not properly re-run the migrations.

**Comment by pmoust** [comment]
It's been a year now. I am keeping this open another week, if there is not much interest it's gone.

**Comment by SleepWalker** [comment]
This is not so hard to implement. The problem is to take the right decision. It can be additional flag for `createTable` or it can be separate method e.g. `createTableIfNotExists`. The second looks better for me, but, probably, not for yii. So we should update function signature to `createTable($table, $columns, $options=null, $ifNotExists = false)`.

If it is acceptable, I can create a pull request

**Comment by nineinchnick** [comment]
I think it should be implemented as a condition that detects `$options` which are to be prepended before table name, not at the end. This way it doesn't break BC and doesn't introduce new arguments. All such options are well defined.

**Comment by pmoust** [comment]
@nineinchnick :+1:

**Comment by SleepWalker** [comment]
@nineinchnick and when I need to use appended `$options` to? Or in this case options should be an array? E.g.:
```php
array(
    'ifNotExists' => true,
    'tableOptions' => 'ENGINE = InnoDb'
);
```

**Comment by nineinchnick** [comment]
Yes, I assumed that $options could be either a string or an numeric array with valid SQL options as values.

**Comment by Horlacher** [comment]
Extended my CDbMigration, maybe this is useful for somebody:
```php
class DbMigration extends CDbMigration
{
    public function tableExists($table)
    {
        return (Yii::app()->db->schema->getTable($table, true) !== null);
    }

    public function dropTableIfExists($table)
    {
        if ($this->tableExists($table)) {
            echo " table $table exists, drop it";
            $this->dropTable($table);
        }
    }

    public function createTable($table, $columns, $options = null, $replace = false)
    {
        if ($replace) {
            $this->dropTableIfExists($table);
        }
        parent::createTable($table, $columns, $options);
    }
}
```

**Comment by pmoust** [comment]
Been over a year, not enough voices heard in this issue so I guess not that many devs see it as a necessity, nor was there a general consensus.

Closing.

**Comment by ognistyi** [comment]
Yii2 approach:
```php
    public function tableExists($tableName)
    {
        return in_array($tableName, Yii::$app->db->schema->tableNames);
    }
```

**Comment by snivs** [comment]
I really wanted to do this for Yii2, so I extended yii\db\Migration and made an implementation that adds any missing columns when the table already exists. I'll leave it here:

    use yii\db\Migration as BaseClass;

    class Migration extends BaseClass {
      public function tableExists($tableName){
        return in_array($tableName, Yii::$app->db->schema->tableNames);
      }

      public function createTable($table, $columns, $options = null) {
        if ($this->tableExists($table)){
          echo "    > table '$table' already exists\n";
          $this->completeColumns($table,$columns);
          return null;
        }
        return parent::createTable($table, $columns, $options);
      }

      public function completeColumns($table,$columns){
        $schema = $this->getDb()->getTableSchema($table);
        /* @var $column string */
        foreach ($columns as $column=>$type){
          if (is_int($column)){continue;}
          if (is_null($schema->getColumn($column))){
            $this->addColumn($table, $column, $type);
            echo "        > column '$column' added\n";
          }
        }    
      }
    }

**Comment by misikch** [comment]
```
use yii\db\Migration;
use yii\db\Schema;

class m160818_170238_create_news_table_users extends Migration
{
    private $tableName = 'users';

    public function up()
    {

        if (in_array($this->tableName, $this->getDb()->schema->tableNames)) {
            $this->dropTable($this->tableName);
        }
        ..... your create table code ....
    }

}
```

**Comment by fbens** [comment]
I'm using the following method instead.
Why doesn't yii2 come with a function like tableExist() ?

```
        if (Yii::$app->db->getTableSchema('{{%m0_link}}', true) === null) {
            printf(" table %s is not exists, continue ... ", $this->db->schema->getRawTableName('{{%m0_link}}'));

            return;
        }

```

**Comment by samdark** [comment]
@fbens you're asking it in a closed issue in Yii 1.1 issue tracker ;)
